src/java/org/apache/cassandra/config/DatabaseDescriptor.java

```diff
@@ -1656,14 +1653,18 @@ public static int getUnloggedBatchAcrossPartitionsWarnThreshold()
 
     public static void setBatchSizeWarnThresholdInKiB(int threshold)
     {
-        DataStorageSpec.IntKibibytesBound storage = new DataStorageSpec.IntKibibytesBound(threshold);
-        checkValidForByteConversion(storage, "batch_size_warn_threshold");
-        conf.batch_size_warn_threshold = new DataStorageSpec.IntKibibytesBound(threshold);
+        conf.batch_size_warn_threshold = createIntKibibyteBoundAndEnsureItIsValidForByteConversion(threshold,"batch_size_warn_threshold");
     }
 
     public static void setBatchSizeFailThresholdInKiB(int threshold)
     {
-        conf.batch_size_fail_threshold = new DataStorageSpec.IntKibibytesBound(threshold);
+        conf.batch_size_fail_threshold = createIntKibibyteBoundAndEnsureItIsValidForByteConversion(threshold,"batch_size_fail_threshold");
+    }
+
+    private static DataStorageSpec.IntKibibytesBound createIntKibibyteBoundAndEnsureItIsValidForByteConversion(int byteValue, String configName){
```

Comment:
Nit: the final brace should be on a new line, K&R style:
```suggestion
    private static DataStorageSpec.IntKibibytesBound createIntKibibyteBoundAndEnsureItIsValidForByteConversion(int byteValue, String configName)
    {
```
Also, the `byteValue` argument is meant to contain a value in kibibytes, so we should probably name it `kibibyteValue`. Or maybe just `kibibytes`, as it's named in the`DataStorageSpec.IntKibibytesBound` constructor that we use immediately below:
```suggestion
    private static DataStorageSpec.IntKibibytesBound createIntKibibyteBoundAndEnsureItIsValidForByteConversion(int kibibytes, String configName)
    {
```
I would also rename the `configName` argument to `propertyName`, I think:
```suggestion
    private static DataStorageSpec.IntKibibytesBound createIntKibibyteBoundAndEnsureItIsValidForByteConversion(int kibibytes, String propertyName)
    {
```
I wonder if we should put this new method and the related `checkValidForByteConversion` closer to each other in the file, wdyt?

One last thought about this is that perhaps we should move this validation into `DataStorageSpec.IntKibibytesBound`, maybe with a validation method or with static constructor performing the validation. But we don't have to do that during this ticket.

Reply:
batch_size_fail_threshold doesn't need the validation and I assumed this part of the patch will be removed.
Anyway, I am not against adding a nice utility method and agree with @adelapena's review comments.

> One last thought about this is that perhaps we should move this validation into DataStorageSpec.IntKibibytesBound, maybe with a validation method or with a static constructor performing the validation. But we don't have to do that during this ticket. 

Didn't we intentionally leave it out of those classes and opted out of custom static methods and constructors as it is a corner case for only a few parameters?  I thought we didn't want to encourage too much customizations in those classes to ensure we keep them clean, only in case we consider this will be a very common case.

Reply:
Thankfully we don't need to handle the precision issue here as going to smaller unit we use multiplication. Overflow is the only issue we need to handle. 
Yes, both` DataStorageSpec.IntKibibytesBound` and `DataStorageSpec.IntMebibytesBound` will need something like `toBytesInLong`.

It seems that `batch_size_fail_threshold` is the only parameter of type `IntKibibytesBound` that suffers from this issue. 
Now, I went ahead and did a further check for the other properties that are widen to long as it seems this bug was introduced when we removed the static methods from `DataStorageSpec`.
We will need to add `toBytesInLong` also for `IntMebibytesBound` to cover ```
getMinFreeSpacePerDriveInBytes() / min_free_space_per_drive,
compaction_large_partition_warning_threshold / getMinFreeSpacePerDriveInBytes(), getMaxHintsFileSize() / max_hints_file_size
```

Reply:
Not sure whether we want to cover all this in this one ticket or split

Reply:
@ekaterinadimitrova2 @adelapena I have a couple of questions

1. Why do we have `DataStorageUnit` as a constructor parameter in subclasses of `DataStorageSpec`? E.g., Isn't `IntKibibytesBound` represents values like 1 KiB, 100 KiB. Why do we have an option like this `new DataStorageSpec.IntKibibytesBound(10, MEBIBYTES)`

2. Shouldn't the `quantity` of the constructor of IntBound classes be int? It is long right now.

3. What is the upper limit of  `batch_size_fail_threshold`?

Either I missed the idea behind `DataStorageSpec`, or the class hierarchy is misleading.

Reply:
This is part of the new config format where users can provide unit of their choice. Now, they cannot provide at the moment any unit smaller than the one used internally by Cassandra for a respective property because during conversion there can be lost precision. Also, we will run into more issues with the accepted upper bound. So in the case we call `DataStorageSpec.IntKibibytesBound(10, MEBIBYTES)` which says - the user wants/provides 10Mebibytes for a property which was type Int and it is used in Kibibytes by Cassandra internally. So it can't be more than Integer.MAX_VALUE-1 in kibibytes. Please refer to the Java Doc and this document - https://cassandra.apache.org/doc/trunk/cassandra/new/configuration.html

`batch_size_fail_threshold `- It can't be more than `Integer.MAX_VALUE` in kibibytes. So the long widening when we return it in bytes can accommodate this. But currently `getBatchSizeFailThreshold()` uses the` IntKibibytesBound.toBytes()` which would return Int and this is a bug. We will need to create a new method `IntKibibytesBound.toBytesInLong()` to be used by `getBatchSizeFailThreshold() `

Reply:
I tried to fix the issues that you mentioned before as well. Please retake a look.